time-series-analyzer-template/src/agent.py

```diff
@@ -0,0 +1,208 @@
+        start_date = end_date - timedelta(minutes=BUCKET_WINDOW_IN_MINUTES * TRAINING_WINDOW_IN_BUCKET_SIZE)
+        logging.info(f"Analyzing alerts from {start_date} to {end_date}")
+
+        # get all alerts for date range
+        df_bot_alerts = forta_explorer.alerts_by_bot(BOT_ID, ALERT_NAME, CONTRACT_ADDRESS, start_date, end_date)
+        logging.info(f"Fetched {len(df_bot_alerts)} for bot_id {BOT_ID}, alert_id {ALERT_NAME}, contract_address {CONTRACT_ADDRESS}")
+
+        if len(df_bot_alerts) == 0:
+            logging.info("No alerts found for bot_id {BOT_ID}, alert_id {ALERT_NAME}, contract_address {CONTRACT_ADDRESS}")
+            MUTEX = False
+            return
+
+        # build time series model without last bucket
+        df_timeseries = df_bot_alerts.resample(str(BUCKET_WINDOW_IN_MINUTES) + 'min', on='createdAt').count()["hash"].reset_index()
+        df_timeseries['createdAt'] = df_timeseries['createdAt'].dt.tz_localize(None)
+
+        if len(df_timeseries) < 3:
+            logging.info("Not enough data to train model")
+            MUTEX = False
+            return
+
+        df_timeseries = df_timeseries[df_timeseries["createdAt"] < df_timeseries["createdAt"].max()]  # this row could be incomplete, so we discard
+        df_current_value = df_timeseries[df_timeseries["createdAt"] == df_timeseries["createdAt"].max()]
+        df_timeseries = df_timeseries[df_timeseries["createdAt"] < df_timeseries["createdAt"].max()]  # this row is what we want to assess against the model, so we discard
+
+        df_timeseries.rename(columns={'createdAt': 'ds', 'hash': 'y'}, inplace=True)
+        df_timeseries['ds'] = df_timeseries['ds'].dt.tz_localize(None)
+
+        # fill in missing values with median
+        median = df_timeseries['y'].median()
+        logging.info(f"Median is {median}.")
+        current_date = start_date - timedelta(minutes=start_date.minute % BUCKET_WINDOW_IN_MINUTES,
+                                              seconds=start_date.second,
+                                              microseconds=start_date.microsecond)
+        current_date += timedelta(minutes=5)
+
+        # first ensure we have values that span start to end date
+        count = 0
+        while(current_date < end_date - timedelta(minutes=BUCKET_WINDOW_IN_MINUTES)):
+            if pd.Timestamp(current_date) not in df_timeseries['ds'].values:
+                count += 1
+                df_timeseries = pd.concat([df_timeseries, pd.DataFrame({'ds': current_date, 'y': median}, index=[df_timeseries.index.max() + 1])])
+            current_date = current_date + timedelta(minutes=BUCKET_WINDOW_IN_MINUTES)
+        logging.info(f"Filled in {count} values.")
+
+        # for any values we do have that are 0, replace with median
+        logging.info(f"Replaced {len(df_timeseries[df_timeseries['y'] == 0])} values with median.")
+        df_timeseries.replace(0, df_timeseries['y'].median(), inplace=True)
```

Comment:
the median is set on line 134, so this could be 
`df_timeseries.replace(0, median, inplace=True)`

Reply:
curious why buckets with 0 alerts are getting replaced with median. Wouldn't 0 alerts be useful signal that the bot is quiet for specific time intervals?

Reply:
This is to deal with outages that I have observed, where it is clear that the alert was just not functioning on the network. If there are quiet alerts, then the median will be 0 and it would replace 0 with 0.

Reply:
ill fix the value as you suggest